subsys/usb/class/netusb/function_ecm.c

```diff
@@ -0,0 +1,297 @@
+		SYS_LOG_DBG("Got frame delimeter, ECM pkt received, len %u",
+			    net_pkt_get_len(in_pkt));
+
+		if (ecm.skip) {
+			SYS_LOG_WRN("End skipping fragments");
+			ecm.skip = false;
+
+			return;
+		}
+
+		net_hexdump_frags(">", in_pkt);
+
+		if (net_recv_data(ecm.iface, in_pkt) < 0) {
+			SYS_LOG_ERR("Queueing packet %p failed", in_pkt);
+			net_pkt_unref(in_pkt);
+		}
+
+		in_pkt = NULL;
+		return;
+	}
+
+	if (ecm.skip) {
+		SYS_LOG_WRN("Skipping %u bytes", len);
+
+		if (len < sizeof(buffer)) {
+			SYS_LOG_WRN("End skipping fragments");
+			ecm.skip = false;
+
+			return;
+		}
+
+		return;
+	}
+
+	/* Start new packet */
+	if (!in_pkt) {
+		pkt = net_pkt_get_reserve_rx(0, K_NO_WAIT);
+		if (!pkt) {
+			SYS_LOG_ERR("Not enough memory for pkt buffer");
+			ecm.skip = true;
+			return;
+		}
+
+		buf = net_pkt_get_frag(pkt, K_NO_WAIT);
+		if (!buf) {
+			net_pkt_unref(pkt);
+			SYS_LOG_ERR("Not enough memory for network buffers");
+			return;
+		}
+
+		net_pkt_frag_insert(pkt, buf);
+
+		in_pkt = pkt;
+	}
+
+	if (!net_pkt_append_all(in_pkt, len, buffer, K_FOREVER)) {
+		SYS_LOG_ERR("Error appending data to pkt: %p", in_pkt);
+
+		net_pkt_unref(in_pkt);
+		in_pkt = NULL;
+		return;
+	}
+
+	if (len < sizeof(buffer)) {
+		SYS_LOG_DBG("ECM network packet received, len %u",
+			    net_pkt_get_len(in_pkt));
+
+		net_hexdump_frags(">", in_pkt);
+
+		if (net_recv_data(ecm.iface, in_pkt) < 0) {
+			SYS_LOG_ERR("Queueing packet %p failed", in_pkt);
+			net_pkt_unref(in_pkt);
+		}
+
+		in_pkt = NULL;
+	}
+}
+
+void ecm_bulk_in(u8_t ep, enum usb_dc_ep_cb_status_code ep_status)
+{
+#if VERBOSE_DEBUG
+	SYS_LOG_DBG("EP 0x%x status %d", ep, ep_status);
+#endif
+}
+
+/*
+ * The idea here is to use one buffer of size endpoint MPS (64 bytes)
+ * for sending pkt without linearlizing first since we would need Ethernet
+ * packet frame as a buffer up to 1518 bytes and it would require two
+ * iterations.
+ */
+static int append_bytes(u8_t *out_buf, u16_t buf_len, u8_t *data,
+			u16_t len, u16_t remaining)
+{
+	int ret;
+
+	do {
+		u16_t count = min(len, remaining);
+#if VERBOSE_DEBUG
+		SYS_LOG_DBG("len %u remaining %u count %u", len, remaining,
+			    count);
+#endif
+
+		memcpy(out_buf + (buf_len - remaining), data, count);
+
+		data += count;
+		remaining -= count;
+		len -= count;
+
+		/* Buffer filled */
+		if (remaining == 0) {
+#if VERBOSE_DEBUG
+			net_hexdump("fragment", out_buf, buf_len);
+#endif
+
+			ret = try_write(ecm.ep_in, out_buf, buf_len);
+			if (ret) {
+				SYS_LOG_ERR("Error sending data");
+				return ret;
+			}
+
+			/* Consumed full buffer */
+			if (len == 0) {
+				return buf_len;
+			}
+
+			remaining = buf_len;
+		}
+	} while (len);
+#if VERBOSE_DEBUG
+	net_hexdump("fragment", out_buf, buf_len - remaining);
+#endif
+	return remaining;
+}
+
+int ecm_send(struct net_pkt *pkt)
+{
+	u8_t send_buf[CONFIG_CDC_ECM_BULK_EP_MPS];
+	int remaining = sizeof(send_buf);
+
+	net_hexdump_frags("<", pkt);
+
+	if (!pkt->frags) {
```

Comment:
If `pkt->frags` is NULL, the `net_pkt_ll()` call below will crash: it calls `net_pkt_ip_data()`, which dereferences `pkt->frags`.  I could fix this, but I don't know the intent of this code here.

Reply:
This is fixed in PR #4641